service/datasync/exec_data_sync.go

```diff
@@ -0,0 +1,238 @@
+	// concurrency safe, so there is no problem with popping from the back while
+	// the poll is pushing new items in the front.
+	record := e.buffer.PopBack()
+	return record.(*uploader.BlockData), nil
+}
+
+func (e *ExecDataSync) poll() {
+
+	// We only call `Next()` sequentially, so there is no need to guard it from
+	// concurrent access. However, when the buffer is not empty, we might still
+	// be polling for new data in the background when the next call happens. We
+	// thus need to ensure that only one poll is executed at the same time. We
+	// do this with a simple flag that is set atomically to work like a
+	// `TryLock()` on a mutex, which is unfortunately not available in Go, see:
+	// https://github.com/golang/go/issues/6123
+	if !atomic.CompareAndSwapUint32(&e.busy, 0, 1) {
+		return
+	}
+	defer atomic.StoreUint32(&e.busy, 0)
+
+	// At this point, we try to pull new files from the cloud.
+	err := e.getExecData()
+	if errors.Is(err, storage.ErrObjectNotExist) {
+		e.log.Debug().Msg("next execution record not available, download stopped")
+		return
+	}
+	if err != nil {
+		e.log.Error().Err(err).Msg("could not download execution records")
+		return
+	}
+}
+
+func (e *ExecDataSync) getExecData() error {
+
+	for {
+
+		// We only want to retrieve and process execData blocks until the buffer is full.
+		if uint(e.buffer.Len()) >= e.limit {
+			e.log.Debug().Uint("limit", e.limit).Msg("buffer full, stopping execution record pull")
+			return nil
+		}
+
+		// We only want to retrieve and process files for blocks that have already
+		// been finalized, in the order that they have been finalized. This
+		// causes some latency, as we don't download until after a block is
+		// finalized, even if the data is available before. However, it seems to
+		// be the only way to make sure trie updates are delivered to the mapper
+		// in the right order without changing the way uploads work.
+		if uint(e.queue.Len()) == 0 {
+			e.log.Debug().Msg("queue empty, stopping execution record download")
+			return nil
+		}
+
+		// Get the name of the file based on the block ID. The file n
+		blockID := e.queue.PopBack().(flow.Identifier)
+		record, err := e.pullData(context.Background(), blockID)
+		if err != nil {
+			e.queue.PushBack(blockID)
+			return fmt.Errorf("could not pull execution record (name: %s): %w", blockID, err)
+		}
+
+		e.log.Debug().
+			Hex("blockID", blockID[:]).
+			Msg("pushing execution record into buffer")
+
+		e.buffer.PushFront(record)
+	}
+}
+
+func (e *ExecDataSync) pullData(ctx context.Context, blockID flow.Identifier) (*uploader.BlockData, error) {
+	// query AN from cached connection
+	req := &execData.GetExecutionDataByBlockIDRequest{BlockId: blockID[:]}
+	res, err := e.execDataApi.GetExecutionDataByBlockID(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return convertToBlockData(res.BlockExecutionData)
+}
+
+func convertToBlockData(data *entities.BlockExecutionData) (*uploader.BlockData, error) {
```

Comment:
let's add a comment here as to why we're doing this conversion instead of using `entites.BlockExecutionData` throughout